unittests/producer_schedule_tests.cpp

```diff
@@ -11,187 +11,122 @@ using mvo = fc::mutable_variant_object;
 
 BOOST_AUTO_TEST_SUITE(producer_schedule_tests)
 
-   // Calculate expected producer given the schedule and slot number
-   account_name get_expected_producer(const vector<producer_authority>& schedule, const uint64_t slot) {
-      const auto& index = (slot % (schedule.size() * config::producer_repetitions)) / config::producer_repetitions;
-      return schedule.at(index).producer_name;
-   };
+namespace {
+
+// Calculate expected producer given the schedule and slot number
+account_name get_expected_producer(const vector<producer_authority>& schedule, block_timestamp_type t) {
+   const auto& index = (t.slot % (schedule.size() * config::producer_repetitions)) / config::producer_repetitions;
+   return schedule.at(index).producer_name;
+};
+
+// Check if two schedule is equal
+bool is_schedule_equal(const vector<producer_authority>& first, const vector<producer_authority>& second) {
+   bool is_equal = first.size() == second.size();
+   for (uint32_t i = 0; i < first.size(); i++) {
+      is_equal = is_equal && first.at(i) == second.at(i);
+   }
+   return is_equal;
+};
+
+} // anonymous namespace
+
+BOOST_FIXTURE_TEST_CASE( verify_producer_schedule, validating_tester ) try {
+
+   // Utility function to ensure that producer schedule work as expected
+   const auto& confirm_schedule_correctness = [&](const vector<producer_authority>& new_prod_schd, uint32_t expected_schd_ver)  {
+      const uint32_t check_duration = 1000; // number of blocks
+      bool scheduled_changed_to_new = false;
+      for (uint32_t i = 0; i < check_duration; ++i) {
+         const auto current_schedule = control->head_block_state()->active_schedule.producers;
+         if (is_schedule_equal(new_prod_schd, current_schedule)) {
+            scheduled_changed_to_new = true;
+         }
+
+         // Produce block
+         produce_block();
 
-   // Check if two schedule is equal
-   bool is_schedule_equal(const vector<producer_authority>& first, const vector<producer_authority>& second) {
-      bool is_equal = first.size() == second.size();
-      for (uint32_t i = 0; i < first.size(); i++) {
-         is_equal = is_equal && first.at(i) == second.at(i);
+         // Check if the producer is the same as what we expect
+         const auto block_time = control->head_block_time();
+         const auto& expected_producer = get_expected_producer(current_schedule, block_time);
+         BOOST_TEST(control->head_block_producer() == expected_producer);
+
+         if (scheduled_changed_to_new)
+            break;
```

Comment:
Shouldn't it be like this:

```suggestion
          if (scheduled_changed_to_new) {
            const auto block_time = control->head_block_time();
            const auto& expected_producer = get_expected_producer(new_prod_schd, block_time);
            BOOST_TEST(control->head_block_producer() == expected_producer);
            break;
         }
```

Reply:
It is good to test that the `expected_producer` is correct even before the schedule change.

Reply:
should you call: get_expected_producer(**new_prod_schd**?

Reply:
`new_prod_schd` is equal to `current_schedule` when the schedule changes, which is why the test was working already.

Reply:
Hum, I guess. The test is a bit strange because the real test is that we actually break out of the loop before the loop condition `i < check_duration` is hit. And doing `BOOST_TEST(control->head_block_producer() == expected_producer);` before `scheduled_changed_to_new` is true does not really check much.
But still a valid test! Approving the PR.